### ISSUE TITLE
Installer tests don't show segfaults since the transition to nix 2.0 and nixos-enter

### ISSUE THREAD

**Issue by samueldr** [NixOS/nixpkgs#46763]
## Issue description

I wasn't able to *conclusively* put the finger on one commit, but I was able to observe that at the merge of the branch `nix-2.0` the segfaults are not present in the installer test logs, while they are in the commit preceding the merge.

 * d586f5d97e8d37f43eae2fe5c98598ce1c393688 "Merge branch 'nix-2.0'"
 * c94ad64ae5573f529213c4d97baf3ea67e86b6b7 "rsbep: init at 0.1.0" (preceding good commit)

This can also be observed on hydra, test failures for grub1 stopped showing the segfault.

> The segfaulting part is `grub`, the "shell" part of grub. On current nixpkgs, it can be shown segfaulting easily > by using `nix-shell -p grub --run grub`.
> 
> ```
> ~/tmp/nixpkgs/nixpkgs-test-segfault $ nix-shell -p grub --run grub
> Probing devices to guess BIOS drives. This may take a long time.
> /tmp/nix-shell-30209-0/rc: line 1: 31572 Segmentation fault      grub
> ```

This isn't the issue here, the issue here is that we probably need to *somehow* get back the segfaults into the logs for the installer tests.

cc @edolstra, as he authored `nixos-enter`. I don't entirely know how it differs to what was previously done to chroot and install, in which it would matter for printing segfaults.

### Steps to reproduce

Run the `grub1` installer tests for each of the previously listed commits.

##### d586f5d97e8d37f43eae2fe5c98598ce1c393688 "Merge branch 'nix-2.0'"

```
 $ nix-build nixos/tests/installer.nix -A grub1
[...]
machine# updating GRUB 1 menu...
machine# installing the GRUB 1 boot loader on /dev/sda...
machine# /nix/store/6bf7yksn9q22p4bqh15ih4bvkmnir3pi-grub-0.97-patch-1.12/sbin/grub-install: line 328: /tmp/grub-install.log.163: No such file or directory
machine# grep: /tmp/grub-install.log.163: No such file or directory
machine# sed: can't read /44V9x5vg6M/boot/grub/device.map: No such file or directory
machine# grep: /44V9x5vg6M/boot/grub/device.map: No such file or directory
machine# /dev/sda does not have any corresponding BIOS drive.
machine# /nix/store/ixvwjckxam1rcvlklw1ap6pb3w0lr58w-install-grub.pl: installation of GRUB on /dev/sda failed
machine: exit status 1
machine: output:
error: command `nixos-install < /dev/null >&2' did not succeed (exit code 1)
[...]
```

##### c94ad64ae5573f529213c4d97baf3ea67e86b6b7 "rsbep: init at 0.1.0" (preceding good commit)

```
 $ nix-build nixos/tests/installer.nix -A grub1
machine# updating GRUB 1 menu...
machine# installing the GRUB 1 boot loader on /dev/sda...
machine# Probing devices to guess BIOS drives. This may take a long time.
machine# [  356.570578] print_req_error: I/O error, dev fd0, sector 0
machine# [  356.571452] floppy: error 10 while reading block 0
machine# [  356.584831] show_signal_msg: 73 callbacks suppressed
machine# [  356.584833] grub[27572]: segfault at f18 ip 0000000008049909 sp 00000000f7a15ffc error 6 in grub[8048000+2a000]
machine# /nix/store/6bf7yksn9q22p4bqh15ih4bvkmnir3pi-grub-0.97-patch-1.12/sbin/grub-install: line 337: 27572 Segmentation fault      $grub_shell --batch $no_floppy --device-map=$device_map  > $log_file <<EOF
machine# quit
machine# EOF
machine#
machine# [  356.686294] grub[27644]: segfault at f18 ip 0000000008049909 sp 00000000f79f1ffc error 6 in grub[8048000+2a000]
machine# /nix/store/6bf7yksn9q22p4bqh15ih4bvkmnir3pi-grub-0.97-patch-1.12/sbin/grub-install: line 411: 27644 Segmentation fault      $grub_shell --batch $no_floppy --device-map=$device_map  > $log_file <<EOF
machine# dump ${root_drive}${tmp} ${img_file}
machine# quit
machine# EOF
machine#
machine# cmp: /tmp/grub-install.img.27565: No such file or directory
machine# [  357.704025] grub[27649]: segfault at f18 ip 0000000008049909 sp 00000000f7971ffc error 6 in grub[8048000+2a000]
machine# /nix/store/6bf7yksn9q22p4bqh15ih4bvkmnir3pi-grub-0.97-patch-1.12/sbin/grub-install: line 411: 27649 Segmentation fault      $grub_shell --batch $no_floppy --device-map=$device_map  > $log_file <<EOF
machine# dump ${root_drive}${tmp} ${img_file}
machine# quit
machine# EOF
machine#
machine# cmp: /tmp/grub-install.img.27565: No such file or directory
machine# [  358.722851] grub[27654]: segfault at f18 ip 0000000008049909 sp 00000000f7994ffc error 6 in grub[8048000+2a000]
machine# /nix/store/6bf7yksn9q22p4bqh15ih4bvkmnir3pi-grub-0.97-patch-1.12/sbin/grub-install: line 411: 27654 Segmentation fault      $grub_shell --batch $no_floppy --device-map=$device_map  > $log_file <<EOF
machine# dump ${root_drive}${tmp} ${img_file}
machine# quit
machine# EOF
machine#
machine# cmp: /tmp/grub-install.img.27565: No such file or directory
machine# [  359.743449] grub[27659]: segfault at f18 ip 0000000008049909 sp 00000000f79dbffc error 6 in grub[8048000+2a000]
machine# /nix/store/6bf7yksn9q22p4bqh15ih4bvkmnir3pi-grub-0.97-patch-1.12/sbin/grub-install: line 411: 27659 Segmentation fault      $grub_shell --batch $no_floppy --device-map=$device_map  > $log_file <<EOF
machine# dump ${root_drive}${tmp} ${img_file}
machine# quit
machine# EOF
machine#
machine# cmp: /tmp/grub-install.img.27565: No such file or directory
machine# [  360.760830] grub[27664]: segfault at f18 ip 0000000008049909 sp 00000000f7974ffc error 6 in grub[8048000+2a000]
machine# /nix/store/6bf7yksn9q22p4bqh15ih4bvkmnir3pi-grub-0.97-patch-1.12/sbin/grub-install: line 411: 27664 Segmentation fault      $grub_shell --batch $no_floppy --device-map=$device_map  > $log_file <<EOF
machine# dump ${root_drive}${tmp} ${img_file}
machine# quit
machine# EOF
machine#
machine# cmp: /tmp/grub-install.img.27565: No such file or directory
machine# The file /tmp/EGzKVZ9VAl/boot/grub/stage1 not read correctly.
machine# /nix/store/ixvwjckxam1rcvlklw1ap6pb3w0lr58w-install-grub.pl: installation of GRUB on /dev/sda failed
machine: exit status 1
machine: output:
error: command `nixos-install < /dev/null >&2' did not succeed (exit code 1)
[...]
```

## Technical details

*This is probably irrelevant, as the installer tests are not affected by this, I believe. See also the previous parts which are pretty technical.*

```
system: "x86_64-linux", multi-user?: yes, version: nix-env (Nix) 2.0.4, channels(samuel): "unstable", channels(root): "nixpkgs-18.09pre144939.14a9ca27e69", nixpkgs: /etc/nixos/nixpkgs
```

**Comment by samueldr** [comment]
Well, I'm probably wrong somewhat? With the alternative *grub1 with fixes but not all of them applied* (when verifying they're all needed) I had segfaults show up on an up-to-date 18.09 checkout on the grub1 installer test.

This is leaving me more confused, and probably makes my previous assumptions partially wrong, if not outright wrong.

Using the fix from #46769, but leaving the gcc6 stdenv out, you can reproduce working segfaults where the previous tests did not :confused:.

```
machine# [  191.601821] show_signal_msg: 145 callbacks suppressed
machine# [  191.601823] grub[29753]: segfault at f38 ip 000000000804983c sp 00000000f7a1a000 error 6 in grub[8048000+2a000]
machine# /nix/store/3z26g0jvcsk7s9nrlri88jk9ash1fbhv-grub-0.97-73/sbin/grub-install: line 386:   167 Segmentation fault      $grub_shell --batch $no_floppy --device-map=$device_map  > $log_file <<EOF
machine# quit
machine# EOF
machine#
machine# /nix/store/3z26g0jvcsk7s9nrlri88jk9ash1fbhv-grub-0.97-73/sbin/grub-install: line 468: grub-probe: command not found
machine# [  191.716081] grub[29832]: segfault at f38 ip 000000000804983c sp 00000000f79aa000 error 6 in grub[8048000+2a000]
machine# /nix/store/3z26g0jvcsk7s9nrlri88jk9ash1fbhv-grub-0.97-73/sbin/grub-install: line 474:   246 Segmentation fault      $grub_shell --batch $no_floppy --device-map=$device_map  > $log_file <<EOF
machine# dump ${root_drive}${tmp} ${img_file}
machine# quit
machine# EOF
machine#
```

**Comment by samueldr** [comment]
??????

What even is this all about?
